File: server/compute/circuit/benches.py
"""Circuit test benches: documented defaults, netlist builders and per-bench analysis.

This module is pure Python/numpy (no numba import), so the API process may import
``BENCH_DEFAULTS`` / ``BENCH_INFO`` / ``SOLVER_DEFAULTS`` for documentation.

All voltages in V, times in s, resistances in ohm, capacitances in F, currents in A
(light: pA, as in the device block).  ``None`` means "automatic" (resolved value returned in
``result["bench_params"]``).

Benches
-------
load_line  V_src(t) triangular v_min -> v_max -> v_min at rate_V_per_s, n_cycles; series R_s;
           drain capacitance C_d; DC gate V_G.  Latch-up / latch-down detected per cycle as I_D
           crossing detect.i_threshold_A upward / downward; V_LU, V_LD at the drain node (device
           V_DS) and at the source (supply).  The body charge Q_B is integrated continuously across
           cycles, so any residual body memory between cycles is included automatically.
pulse      supply pulse train (v_base, v_amp, width = flat top, period, rise, fall, n_pulses)
           through R_s; P_sw = fraction of pulses latched (I_D >= i_threshold) at the end of the flat
           top; also the fraction still latched at the end of the period (retention) and the
           switching delay.  Optional amplitude list -> sweep of P_sw vs amplitude.
pbit       STL + load resistor R_L from a clocked supply; comparator on v_D with threshold
           (default v_high - R_L * 100 nA): bit = 1 when v_D < v_th at the end of the clock high
           phase (cell latched).  Statistics P(1) and lag-1 autocorrelation; optional list of V_G or
           light values -> sweep of P(1).
coupled    two STL cells, each with its own series R to a common ramp or pulse source, coupled by R_c
           between the drains; the second cell may override V_G / light.
"""
from __future__ import annotations

import copy
from typing import Any

import numpy as np

from .netlist import Netlist, pulse_train, triangle

BENCH_INFO = {
    "load_line": {"ko": "부하선 스윕 (직렬 저항 + 삼각파)", "en": "Load-line sweep (series R, triangular ramp)"},
    "pulse": {"ko": "펄스 열 (스위칭 확률 P_sw)", "en": "Pulse train (switching probability P_sw)"},
    "pbit": {"ko": "p-비트 (부하 저항 + 클럭 + 비교기)", "en": "p-bit (load resistor, clock, comparator)"},
    "coupled": {"ko": "저항 결합 STL 두 개", "en": "Two resistively coupled STLs"},
}

BENCH_DEFAULTS: dict[str, dict[str, Any]] = {
    "load_line": dict(
        v_min_V=0.0,            # ramp start/end voltage
        v_max_V=None,           # auto: device preset sweep vd_max (paper 4 V, photo 5 V)
        rate_V_per_s=None,      # auto: preset sweep rate (paper 0.4 V/s, photo 1200 V/s); stochastic
                                #       event-level runs fall back to 1200 V/s when the preset rate is too slow
        n_cycles=1,             # triangular cycles (<= 50)
        R_s_ohm=1e3,            # series resistor
        C_d_F=2e-15,            # drain-node capacitance to ground
        vg_V=None,              # gate voltage (default: device.vg)
    ),
    "pulse": dict(
        v_base_V=0.0,
        v_amp_V=None,           # auto: deterministic latch-up fold V_LU of the device + 0.10 V
        width_s=200e-6,         # flat top
        period_s=1e-3,
        rise_s=1e-6,
        fall_s=1e-6,
        n_pulses=10,            # <= 2000
        delay_s=0.0,            # start of the first pulse
        R_s_ohm=1e3,
        C_d_F=2e-15,
        vg_V=None,
        amplitudes_V=[],        # optional sweep of v_amp (P_sw vs amplitude)
    ),
    "pbit": dict(
        v_low_V=0.0,
        v_high_V=None,          # auto: fold V_LU - 0.02 V (stochastic switching regime)
        clock_period_s=1e-3,
        clock_width_s=200e-6,   # flat top of the clock
        rise_s=1e-6,
        fall_s=1e-6,
        n_clocks=50,            # <= 5000
        R_L_ohm=100e3,
        C_d_F=2e-15,
        cmp_threshold_V=None,   # auto: v_high - R_L * 100 nA
        vg_V=None,
        vg_list_V=[],           # optional sweep of V_G -> P(1)
        light_list_pA=[],       # optional sweep of I_PH -> P(1) (used when vg_list_V is empty)
    ),
    "coupled": dict(
        source="ramp",          # "ramp" | "pulse"
        v_min_V=0.0, v_max_V=None, rate_V_per_s=None, n_cycles=1,               # ramp source
        v_base_V=0.0, v_amp_V=None, width_s=200e-6, period_s=1e-3, rise_s=1e-6,  # pulse source
        fall_s=1e-6, n_pulses=10, delay_s=0.0,
        R_s1_ohm=100e3, R_s2_ohm=100e3,     # series resistors of cell 1 / cell 2
        R_c_ohm=1e6,                        # coupling resistor between the drains
        C_d_F=2e-15,                        # each drain node
        vg_V=None,                          # cell 1 gate (default device.vg)
        vg2_V=None,                         # cell 2 gate (default = cell 1)
        iph2_pA=None,                       # cell 2 photocurrent (default = device light)
    ),
}

SOLVER_DEFAULTS = dict(
    method="BE",                # "BE" | "TRAP" (TRAP falls back to BE on stiff steps h > 2 tau_rel)
    dt_min_s=None,              # auto: max(1e-15, 1e-13 * t_end)
    dt_max_s=None,              # auto: t_end / 2000
    reltol=1e-3,                # scales the per-step limits: |du| <= 10 mV, |dln I| <= 0.2, |dv| <= 20 mV,
                                # LTE(u) <= 1 mV at reltol = 1e-3 (linear in reltol, clipped)
    max_steps=1_000_000,        # per run (cap 2e6)
    tau_frac=0.05,              # stochastic: h <= tau_frac * tau_rel
    max_events_per_step=200,    # stochastic: expected events per step
    noise_dt_min_s=2e-9,        # stochastic: carrier noise resolved where tau_frac*tau_rel >= this; drift-only elsewhere
    gauss_threshold=100,        # stochastic: Poisson counts with mean > this use the Gaussian limit
)

STOCHASTIC_DEFAULTS = dict(seed=2026092920, n_runs=20, carrier_noise=True, local_state=dict(mode="none"))

CAPS = dict(max_steps=2_000_000, n_runs=200, n_cycles=50, n_pulses=2000, n_clocks=5000, sweep_points=25,
            v_abs_max=8.0)


def merged(base: dict, over: dict | None) -> dict:
    out = copy.deepcopy(base)
    for k, v in (over or {}).items():
        if isinstance(v, dict) and isinstance(out.get(k), dict):
            out[k] = merged(out[k], v)
        else:
            out[k] = v
    return out


def _num(d: dict, key: str, lo: float | None = None, hi: float | None = None, positive=False) -> float:
    try:
        v = float(d[key])
    except (TypeError, ValueError):
        raise ValueError(f"bench_params.{key} must be a number")
    if not np.isfinite(v):
        raise ValueError(f"bench_params.{key} must be finite")
    if positive and v <= 0:
        raise ValueError(f"bench_params.{key} must be > 0")
    if lo is not None and v < lo:
        raise ValueError(f"bench_params.{key} must be >= {lo}")
    if hi is not None and v > hi:
        raise ValueError(f"bench_params.{key} must be <= {hi}")
    return v


def _int(d: dict, key: str, lo: int, hi: int, warnings: list[str]) -> int:
    try:
        v = int(round(float(d[key])))
    except (TypeError, ValueError):
        raise ValueError(f"bench_params.{key} must be an integer")
    if v < lo:
        raise ValueError(f"bench_params.{key} must be >= {lo}")
    if v > hi:
        warnings.append(f"bench_params.{key} = {v} capped at {hi}")
        v = hi
    d[key] = v
    return v


def _list(d: dict, key: str, warnings: list[str], lo=None, hi=None) -> list[float]:
    raw = d.get(key) or []
    if not isinstance(raw, (list, tuple)):
        raise ValueError(f"bench_params.{key} must be a list of numbers")
    vals = []
    for x in raw:
        try:
            v = float(x)
        except (TypeError, ValueError):
            raise ValueError(f"bench_params.{key} must contain numbers")
        if not np.isfinite(v) or (lo is not None and v < lo) or (hi is not None and v > hi):
            raise ValueError(f"bench_params.{key}: value {x} out of range [{lo}, {hi}]")
        vals.append(v)
    if len(vals) > CAPS["sweep_points"]:
        warnings.append(f"bench_params.{key}: only the first {CAPS['sweep_points']} values are used")
        vals = vals[:CAPS["sweep_points"]]
    d[key] = vals
    return vals


# ---- builders -------------------------------------------------------------------------------
class BenchSpec:
    """A built bench: netlist + everything the analysis needs."""

    def __init__(self, bench: str, net: Netlist, main_wave: int, meta: dict):
        self.bench = bench
        self.net = net
        self.main_wave = main_wave
        self.meta = meta


def _add_cell(net: Netlist, k: int, drain: str, gate: str, p, vg: float, label: str):
    net.add_V(f"VG{k}", gate, "0", [0.0], [vg], f"{vg:g} V (DC)")
    net.add_STL(f"X{k}", drain, gate, "0", p, label=label)


def build_load_line(bp: dict, p, vg: float, label: str) -> BenchSpec:
    net = Netlist()
    t, v, Tc = triangle(bp["v_min_V"], bp["v_max_V"], bp["rate_V_per_s"], bp["n_cycles"])
    w = net.add_V("Vsrc", "src", "0", t, v, f"triangle {bp['v_min_V']:g}→{bp['v_max_V']:g} V, {bp['rate_V_per_s']:g} V/s")
    net.add_R("Rs", "src", "d", bp["R_s_ohm"])
    net.add_C("Cd", "d", "0", bp["C_d_F"])
    _add_cell(net, 1, "d", "g", p, vg, label)
    net.t_end = float(t[-1])
    net.main_wave = w
    return BenchSpec("load_line", net, w, dict(cycle_s=Tc, n_cycles=bp["n_cycles"], cells=[dict(drain="d", name="X1")]))


def _pulse_wave(bp: dict, amp: float):
    return pulse_train(bp["v_base_V"], amp, bp["width_s"], bp["period_s"], bp["rise_s"], bp["fall_s"],
                       bp["n_pulses"], bp["delay_s"])


def build_pulse(bp: dict, p, vg: float, label: str, amp: float | None = None) -> BenchSpec:
    net = Netlist()
    a = bp["v_amp_V"] if amp is None else amp
    t, v, top_end, per_end = _pulse_wave(bp, a)
    w = net.add_V("Vsrc", "src", "0", t, v, f"pulses {bp['v_base_V']:g}→{a:g} V, {bp['width_s']:.3g} s / {bp['period_s']:.3g} s")
    net.add_R("Rs", "src", "d", bp["R_s_ohm"])
    net.add_C("Cd", "d", "0", bp["C_d_F"])
    _add_cell(net, 1, "d", "g", p, vg, label)
    # sample just before the fall (end of the flat top) and at the end of each period
    per_sample = [te - 1e-6 * bp["period_s"] for te in per_end]
    net.samples = sorted(set(top_end) | set(per_sample))
    net.t_end = float(t[-1])
    net.main_wave = w
    starts = [bp["delay_s"] + i * bp["period_s"] for i in range(bp["n_pulses"])]
    return BenchSpec("pulse", net, w, dict(top_end=top_end, per_sample=per_sample, starts=starts, amp=a,
                                           cells=[dict(drain="d", name="X1")]))


def build_pbit(bp: dict, p, vg: float, label: str) -> BenchSpec:
    net = Netlist()
    t, v, top_end, per_end = pulse_train(bp["v_low_V"], bp["v_high_V"], bp["clock_width_s"], bp["clock_period_s"],
                                         bp["rise_s"], bp["fall_s"], bp["n_clocks"], 0.0)
    w = net.add_V("Vclk", "clk", "0", t, v, f"clock {bp['v_low_V']:g}/{bp['v_high_V']:g} V, {bp['clock_period_s']:.3g} s")
    net.add_R("RL", "clk", "d", bp["R_L_ohm"])
    net.add_C("Cd", "d", "0", bp["C_d_F"])
    _add_cell(net, 1, "d", "g", p, vg, label)
    net.add_CMP("CMP", "d", bp["cmp_threshold_V"], f"bit = [v_D < {bp['cmp_threshold_V']:.4g} V] at clock-high end")
    net.samples = list(top_end)
    net.t_end = float(t[-1])
    net.main_wave = w
    return BenchSpec("pbit", net, w, dict(top_end=top_end, v_th=bp["cmp_threshold_V"], cells=[dict(drain="d", name="X1")]))


def build_coupled(bp: dict, p1, p2, vg1: float, vg2: float, label1: str, label2: str) -> BenchSpec:
    net = Netlist()
    meta: dict[str, Any] = dict(source=bp["source"])
    if bp["source"] == "ramp":
        t, v, Tc = triangle(bp["v_min_V"], bp["v_max_V"], bp["rate_V_per_s"], bp["n_cycles"])
        lab = f"triangle {bp['v_min_V']:g}→{bp['v_max_V']:g} V, {bp['rate_V_per_s']:g} V/s"
        meta.update(cycle_s=Tc, n_cycles=bp["n_cycles"])
    else:
        t, v, top_end, per_end = _pulse_wave(bp, bp["v_amp_V"])
        lab = f"pulses {bp['v_base_V']:g}→{bp['v_amp_V']:g} V"
        net.samples = list(top_end)
        meta.update(top_end=top_end, starts=[bp["delay_s"] + i * bp["period_s"] for i in range(bp["n_pulses"])])
    w = net.add_V("Vsrc", "src", "0", t, v, lab)
    net.add_R("Rs1", "src", "d1", bp["R_s1_ohm"])
    net.add_R("Rs2", "src", "d2", bp["R_s2_ohm"])
    net.add_R("Rc", "d1", "d2", bp["R_c_ohm"])
    net.add_C("Cd1", "d1", "0", bp["C_d_F"])
    net.add_C("Cd2", "d2", "0", bp["C_d_F"])
    _add_cell(net, 1, "d1", "g1", p1, vg1, label1)
    _add_cell(net, 2, "d2", "g2", p2, vg2, label2)
    net.t_end = float(t[-1])
    net.main_wave = w
    meta["cells"] = [dict(drain="d1", name="X1"), dict(drain="d2", name="X2")]
    return BenchSpec("coupled", net, w, meta)


# ---- statistics helpers ---------------------------------------------------------------------
def stats(values) -> dict:
    x = np.asarray([np.nan if v is None else v for v in values], float)
    fin = x[np.isfinite(x)]
    out = dict(n=int(len(fin)), mean=None, sd=None, median=None, p05=None, p95=None, min=None, max=None,
               censored=int(np.sum(~np.isfinite(x))), lag1=None)
    if len(fin):
        out.update(mean=float(fin.mean()), median=float(np.median(fin)), p05=float(np.quantile(fin, 0.05)),
                   p95=float(np.quantile(fin, 0.95)), min=float(fin.min()), max=float(fin.max()))
    if len(fin) > 1:
        out["sd"] = float(fin.std(ddof=1))
    if len(x) > 2:
        a, b = x[:-1], x[1:]
        ok = np.isfinite(a) & np.isfinite(b)
        if ok.sum() > 2 and np.std(a[ok]) > 0 and np.std(b[ok]) > 0:
            out["lag1"] = float(np.corrcoef(a[ok], b[ok])[0, 1])
    return out


def lag1_bits(bits: np.ndarray) -> float | None:
    b = np.asarray(bits, float)
    if len(b) < 3 or np.std(b[:-1]) == 0 or np.std(b[1:]) == 0:
        return None
    return float(np.corrcoef(b[:-1], b[1:])[0, 1])

File: STL-simulator/web/src/params/benches.ts
// Circuit test benches (docs/WEB_CONTRACT.md §4). Keys and defaults mirror
// server/compute/circuit/benches.py (BENCH_DEFAULTS, SOLVER_DEFAULTS, STOCHASTIC_DEFAULTS).
// `null` = "auto" (resolved server-side from the device, e.g. v_max from the preset sweep, v_amp from V_LU, or
// the server's documented default such as rise_s/fall_s). Null keys are omitted from the payload
// (utils/payload.ts circuitPayload) so server/compute/circuit/benches.py BENCH_DEFAULTS applies.
import type { L10n } from "../content/physics/types";
import type { BenchId, CircuitStochBlock, SolverBlock } from "../api/types";
import type { StrKey } from "../i18n/strings";

export type BenchValue = number | string | boolean | null | number[];

export interface BenchField {
  key: string;
  sym?: string;
  label: L10n;
  help: L10n;
  unit: string;
  scale?: number;
  min?: number;
  max?: number;
  step?: number;
  slider?: boolean | "log";
  int?: boolean;
  /** null allowed = automatic value chosen by the server. */
  auto?: boolean;
  type?: "number" | "list" | "select";
  options?: { value: string; label: L10n }[];
  /** Only shown when this predicate on the bench params holds. */
  when?: (bp: Record<string, BenchValue>) => boolean;
  /** List fields: an example in this field's units. */
  placeholder?: L10n;
  /** Auto value the client can resolve (shown as "자동 = …"): the device V_G (server: device.vg), or the
   *  preset's sweep peak / rate (server: the preset's sweep section, not the edited sidebar sweep). */
  autoFrom?: "vg" | "vd_max" | "rate" | "edge" | "vref";
}
export interface BenchDef {
  id: BenchId;
  title: StrKey;
  desc: StrKey;
  defaults: Record<string, BenchValue>;
  fields: BenchField[];
}

const L = (ko: string, en: string): L10n => ({ ko, en });

const f = {
  vMin: { key: "v_min_V", sym: "V_{\\min}", label: L("램프 시작/끝 전압", "Ramp start/end"), help: L("삼각파 최저 전압", "Triangle low level"), unit: "V", min: -8, max: 8, step: 0.05 },
  vMax: { key: "v_max_V", sym: "V_{\\max}", label: L("램프 최대 전압", "Ramp peak"), help: L("자동: 프리셋의 스윕 V_D,max (기준 보정 4 V, 광조사 보정 5 V)", "Auto: the preset's sweep V_D,max (reference 4 V, illumination 5 V)"), unit: "V", min: -8, max: 8, step: 0.05, auto: true, autoFrom: "vd_max" as const },
  rate: { key: "rate_V_per_s", sym: "\\dot V", label: L("램프 속도", "Ramp rate"), help: L("자동: 프리셋 스윕 속도 (확률 실행에서 너무 느리면 1200 V/s)", "auto: preset sweep rate (stochastic runs fall back to 1200 V/s when too slow)"), unit: "V/s", min: 1e-4, max: 1e8, slider: "log" as const, auto: true, autoFrom: "rate" as const },
  nCycles: { key: "n_cycles", sym: "N_{\\mathrm{cyc}}", label: L("삼각파 사이클", "Triangle cycles"), help: L("≤ 50", "≤ 50"), unit: "", min: 1, max: 50, step: 1, int: true },
  Rs: { key: "R_s_ohm", sym: "R_s", label: L("직렬 저항", "Series resistor"), help: L("전원과 드레인 사이 저항", "Resistor between source and drain"), unit: "kΩ", scale: 1e-3, min: 1e-6, max: 1e9, slider: "log" as const },
  Cd: { key: "C_d_F", sym: "C_d", label: L("드레인 노드 용량", "Drain-node capacitance"), help: L("드레인 노드의 접지 용량", "Drain-node capacitance to ground"), unit: "fF", scale: 1e15, min: 0, max: 1e9, slider: "log" as const },
  vg: { key: "vg_V", sym: "V_G", label: L("벤치 게이트 전압", "Bench gate voltage"), help: L("자동: 바이어스·스윕의 V_G", "auto: the V_G of Bias & sweep"), unit: "V", min: -6, max: 1, step: 0.01, auto: true, autoFrom: "vg" as const },
  vBase: { key: "v_base_V", sym: "V_{\\mathrm{base}}", label: L("기저 전압", "Base level"), help: L("펄스 사이 전압", "Voltage between pulses"), unit: "V", min: -8, max: 8, step: 0.05 },
  vAmp: { key: "v_amp_V", sym: "V_{\\mathrm{amp}}", label: L("펄스 진폭", "Pulse amplitude"), help: L("자동: 결정론 V_LU + 0.10 V", "auto: deterministic V_LU + 0.10 V"), unit: "V", min: -8, max: 8, step: 0.01, auto: true },
  width: { key: "width_s", sym: "t_w", label: L("펄스 폭 (평탄부)", "Pulse width (flat top)"), help: L("high 구간", "High-level duration"), unit: "µs", scale: 1e6, min: 1e-6, max: 1e12, slider: "log" as const },
  period: { key: "period_s", sym: "T", label: L("펄스 주기", "Pulse period"), help: L("펄스 반복 주기", "Repetition period"), unit: "µs", scale: 1e6, min: 1e-6, max: 1e12, slider: "log" as const },
  // rise/fall default to the server's values (pulse/coupled 10 µs, p-bit 20 µs): edges faster than ~1 µs kick
  // the floating body through the drain-depletion charge and trigger latch-up below the static fold.
  rise: {
    key: "rise_s", sym: "t_r", label: L("상승 시간", "Rise time"),
    help: L("자동: 서버 기본값 (펄스·결합 10 µs, p-bit 20 µs). ~1 µs보다 빠른 에지는 드레인 공핍 전하를 통해 떠 있는 바디를 차서 정적 fold 아래에서도 래치업을 일으킨다", "auto: server default (pulse/coupled 10 µs, p-bit 20 µs). Edges faster than ~1 µs kick the floating body through the drain-depletion charge and trigger latch-up below the static fold"),
    unit: "µs", scale: 1e6, min: 0, max: 1e9, slider: "log" as const, auto: true, autoFrom: "edge" as const,
  },
  fall: {
    key: "fall_s", sym: "t_f", label: L("하강 시간", "Fall time"),
    help: L("자동: 서버 기본값 (펄스·결합 10 µs, p-bit 20 µs). 빠른 에지는 드레인 공핍 결합으로 바디 전하를 흔든다", "auto: server default (pulse/coupled 10 µs, p-bit 20 µs). Fast edges disturb the body charge through the drain-depletion coupling"),
    unit: "µs", scale: 1e6, min: 0, max: 1e9, slider: "log" as const, auto: true, autoFrom: "edge" as const,
  },
  nPulses: { key: "n_pulses", sym: "N_p", label: L("펄스 수", "Pulses"), help: L("≤ 2000", "≤ 2000"), unit: "", min: 1, max: 2000, step: 1, int: true },
  delay: { key: "delay_s", sym: "t_0", label: L("첫 펄스 지연", "First-pulse delay"), help: L("첫 펄스 시작 시각", "Start of the first pulse"), unit: "µs", scale: 1e6, min: 0, max: 1e12 },
};

export const BENCHES: Record<BenchId, BenchDef> = {
  load_line: {
    id: "load_line",
    title: "c.bench.load_line",
    desc: "c.bench.load_line.desc",
    defaults: { v_min_V: 0, v_max_V: null, rate_V_per_s: null, n_cycles: 1, R_s_ohm: 1e3, C_d_F: 2e-15, vg_V: null },
    fields: [f.vMin, f.vMax, f.rate, f.nCycles, f.Rs, f.Cd, f.vg],
  },
  pulse: {
    id: "pulse",
    title: "c.bench.pulse",
    desc: "c.bench.pulse.desc",
    defaults: { v_base_V: 0, v_amp_V: null, width_s: 200e-6, period_s: 1e-3, rise_s: null, fall_s: null, n_pulses: 10, delay_s: 0, R_s_ohm: 1e3, C_d_F: 2e-15, vg_V: null, amplitudes_V: [] },
    fields: [
      f.vBase, f.vAmp, f.width, f.period, f.rise, f.fall, f.nPulses, f.delay, f.Rs, f.Cd, f.vg,
      { key: "amplitudes_V", sym: "\\{V_{\\mathrm{amp}}\\}", label: L("진폭 스윕 (선택)", "Amplitude sweep (optional)"), help: L("쉼표로 구분한 진폭 목록 → P_sw vs 진폭 (≤ 25개)", "Comma-separated amplitudes → P_sw vs amplitude (≤ 25)"), unit: "V", type: "list", min: -8, max: 8, placeholder: L("예: 3.6, 3.8, 4.0 (비우면 없음)", "e.g. 3.6, 3.8, 4.0 (leave empty for none)") },
    ],
  },
  pbit: {
    id: "pbit",
    title: "c.bench.pbit",
    desc: "c.bench.pbit.desc",
    defaults: { v_low_V: 0, v_high_V: null, clock_period_s: 1e-3, clock_width_s: 200e-6, rise_s: null, fall_s: null, n_clocks: 50, R_S_ohm: 100e3, v_ref_V: null, vg_V: null, vg_list_V: [], light_list_pA: [] },
    fields: [
      { key: "v_low_V", sym: "V_{\\mathrm{low}}", label: L("펄스 low", "Pulse low"), help: L("드레인 펄스의 low 전압", "Low level of the drain pulses"), unit: "V", min: -8, max: 8, step: 0.05 },
      { key: "v_high_V", sym: "V_{\\mathrm{high}}", label: L("펄스 high", "Pulse high"), help: L("자동: V_LU − 15 mV (확률 스위칭 영역, 기준 소자에서 P(1) ≈ 0.5; 결정론 모드에서는 0)", "auto: V_LU − 15 mV (stochastic switching regime: P(1) ≈ 0.5 for the reference device; 0 deterministically)"), unit: "V", min: -8, max: 8, step: 0.01, auto: true },
      { key: "clock_period_s", sym: "T_{\\mathrm{clk}}", label: L("펄스 주기", "Pulse period"), help: L("비트 한 개당 시간", "Time per bit"), unit: "µs", scale: 1e6, min: 1e-6, max: 1e12, slider: "log" },
      { key: "clock_width_s", sym: "t_{\\mathrm{clk}}", label: L("펄스 폭", "Pulse width"), help: L("high 평탄부", "Flat top of the pulse"), unit: "µs", scale: 1e6, min: 1e-6, max: 1e12, slider: "log" },
      f.rise, f.fall,
      { key: "n_clocks", sym: "N_{\\mathrm{clk}}", label: L("비트 수", "Pulses (bits)"), help: L("≤ 5000", "≤ 5000"), unit: "", min: 1, max: 5000, step: 1, int: true, slider: "log" },
      { key: "R_S_ohm", sym: "R_S", label: L("소스 저항", "Source resistor"), help: L("소스와 접지 사이. 래치되면 V(R_S) = R_S·I_D ≈ 0.4 V (100 kΩ)", "Between the source and ground. Latched: V(R_S) = R_S·I_D ≈ 0.4 V (100 kΩ)"), unit: "kΩ", scale: 1e-3, min: 1e-6, max: 1e6, slider: "log" },
      { key: "v_ref_V", sym: "V_{\\mathrm{ref}}", label: L("비교기 기준 전압", "Comparator V_ref"), help: L("자동: R_S × 1 µA. 비트 = [V(R_S) > V_ref], 펄스 평탄부 끝에서", "auto: R_S × 1 µA. bit = [V(R_S) > V_ref] at the end of the pulse flat top"), unit: "V", min: -8, max: 8, step: 0.01, auto: true, autoFrom: "vref" },
      f.vg,
      { key: "vg_list_V", sym: "\\{V_G\\}", label: L("V_G 스윕 (선택)", "V_G sweep (optional)"), help: L("쉼표로 구분 → P(1) vs V_G", "Comma-separated → P(1) vs V_G"), unit: "V", type: "list", min: -6, max: 1, placeholder: L("예: −2.2, −2.0, −1.8 (비우면 없음)", "e.g. −2.2, −2.0, −1.8 (leave empty for none)") },
      { key: "light_list_pA", sym: "\\{I_{PH}\\}", label: L("광전류 스윕 (선택)", "Light sweep (optional)"), help: L("V_G 목록이 비었을 때 사용 → P(1) vs I_PH", "Used when the V_G list is empty → P(1) vs I_PH"), unit: "pA", type: "list", min: 0, max: 1e6, placeholder: L("예: 0, 0.5, 1, 2 (비우면 없음)", "e.g. 0, 0.5, 1, 2 (leave empty for none)") },
    ],
  },
  coupled: {
    id: "coupled",
    title: "c.bench.coupled",
    desc: "c.bench.coupled.desc",
    defaults: {
      source: "ramp", v_min_V: 0, v_max_V: null, rate_V_per_s: null, n_cycles: 1, v_base_V: 0, v_amp_V: null, width_s: 200e-6, period_s: 1e-3, rise_s: null, fall_s: null,
      n_pulses: 10, delay_s: 0, R_s1_ohm: 100e3, R_s2_ohm: 100e3, R_c_ohm: 1e6, C_d_F: 2e-15, vg_V: null, vg2_V: null, iph2_pA: null,
    },
    fields: [
      { key: "source", label: L("공통 전원", "Common source"), help: L("삼각 램프 또는 펄스 열", "Triangular ramp or pulse train"), unit: "", type: "select", options: [{ value: "ramp", label: L("램프", "ramp") }, { value: "pulse", label: L("펄스", "pulse") }] },
      ...[f.vMin, f.vMax, f.rate, f.nCycles].map((x) => ({ ...x, when: (bp: Record<string, BenchValue>) => bp.source !== "pulse" })),
      ...[f.vBase, f.vAmp, f.width, f.period, f.rise, f.fall, f.nPulses, f.delay].map((x) => ({ ...x, when: (bp: Record<string, BenchValue>) => bp.source === "pulse" })),
      { ...f.Rs, key: "R_s1_ohm", sym: "R_{s1}", label: L("직렬 저항 (셀 1)", "Series resistor (cell 1)") },
      { ...f.Rs, key: "R_s2_ohm", sym: "R_{s2}", label: L("직렬 저항 (셀 2)", "Series resistor (cell 2)") },
      { key: "R_c_ohm", sym: "R_c", label: L("결합 저항", "Coupling resistor"), help: L("두 드레인 사이 저항", "Resistor between the drains"), unit: "kΩ", scale: 1e-3, min: 1e-6, max: 1e12, slider: "log" },
      f.Cd,
      { ...f.vg, key: "vg_V", sym: "V_{G,1}", label: L("셀 1 게이트", "Cell 1 gate") },
      { ...f.vg, key: "vg2_V", sym: "V_{G,2}", label: L("셀 2 게이트", "Cell 2 gate"), help: L("자동: 셀 1과 동일", "auto: same as cell 1") },
      { key: "iph2_pA", sym: "I_{PH,2}", label: L("셀 2 광전류", "Cell 2 photocurrent"), help: L("자동: 소자 광 설정", "auto: device light"), unit: "pA", min: 0, max: 1e6, step: 0.01, auto: true },
    ],
  },
};

export const BENCH_ORDER: BenchId[] = ["load_line", "pulse", "pbit", "coupled"];

export const DEFAULT_SOLVER: SolverBlock = {
  method: "BE",
  dt_min_s: null,
  dt_max_s: null,
  reltol: 1e-3,
  max_steps: 1_000_000,
  tau_frac: 0.05,
  max_events_per_step: 200,
  noise_dt_min_s: 2e-9,
  gauss_threshold: 100,
};

export const DEFAULT_CIRCUIT_STOCH: CircuitStochBlock = {
  seed: 2026092920,
  n_runs: 20,
  carrier_noise: true,
  local_state: { mode: "none", action: "gidl", sigma: 0.15339035678526572, tau_s: 5, sigma_E_V: 0.00043696121025623943, tau_E_s: 1.6205564347468981, acquisition_trend: false },
};
